tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11096,6 +11238,1067 @@ TEST_F(ExecutionTest, PackUnpackTest) {
+        VERIFY_IS_TRUE(TestConfig.IntrinsicString != "", "Expecting intrinsic string");
+        if (TestConfig.IntrinsicString == "min")
+          VecExpected.push_back(std::min<T>(Input1, Input2));
+        else if (TestConfig.IntrinsicString == "max")
+          VecExpected.push_back(std::max<T>(Input1, Input2));
+        else
+          LogErrorFmt(L"Unrecognized BinaryOp intrinsic string: %s", TestConfig.IntrinsicString.c_str());
+      } else
+        LogErrorFmt(L"Don't know how to compute expected value for operatorString: %s", TestConfig.OperatorString.c_str());
+    }
+    else // Unary op logic
+    {
+      if (TestConfig.IntrinsicString == "testClamp")
+        VecExpected.push_back(std::clamp(VecInput1[Index], ClampArgC, ClampArgT));
+      else if (TestConfig.IntrinsicString == "testInitialize")
+        VecExpected.push_back(VecInput1[Index]);
+      else
+        LogErrorFmt(L"Unrecognized intrinsic string: %s", TestConfig.IntrinsicString.c_str());
+    }
+  }
+
+  // 1,2, and 3 are required to compile the shader.
+  // 1. -DOPERATOR : "*" "+" "," "-" "/" etc.
+  // 2. -DOPERAND2: : l.VecInput2 or scalarInput, depending on the test.
+  // 3. -DCLAMP_ARGC, -DCLAMP_ARG2 : clamp arg 1 and 2. Required to compile.
+  // 4. Optional: -DFUNC : min, max, etc.
+  std::string ShaderText = R"(
+  struct LongVectorOp{
+    TYPE ScalarInput;
+    vector<TYPE, NUM> VecInput1;
+    vector<TYPE, NUM> VecInput2;
+    vector<TYPE, NUM> VecOutput;
+  };
+
+  vector<TYPE, NUM> testInitialize(vector<TYPE, NUM> Vec)
+  {
+    vector<TYPE, NUM> VecCopy = Vec;
+    return VecCopy;
+  }
+
+  vector<TYPE, NUM> testClamp(vector<TYPE, NUM> Vec)
+  {
+    return clamp(Vec, TYPE(CLAMP_ARGC), TYPE(CLAMP_ARGT));
+  }
+
+  RWStructuredBuffer<LongVectorOp> g_buf : register(u0);
+  [numthreads(8,8,1)]
+  void main(uint GI : SV_GroupIndex) {
+    LongVectorOp LVOp = g_buf[GI];
+    LVOp.VecOutput = FUNC(LVOp.VecInput1 OPERATOR OPERAND2);
+    g_buf[GI] = LVOp;
+  };
+  )";
+
+  // Set up the compiler options string.
+  std::stringstream CompilerOptions("");
+  std::string HLSLType = TestConfig.GetHLSLTypeString();
+  CompilerOptions << "-DTYPE=";
+  CompilerOptions << HLSLType;
+  CompilerOptions << " -DNUM=";
+  CompilerOptions << N;
+  const bool Is16BitType = (HLSLType == "int16_t" || HLSLType == "uint16_t" || HLSLType == "half");
+  CompilerOptions << (Is16BitType ? " -enable-16bit-types" : "");
+  CompilerOptions << " -DOPERATOR=";
+  CompilerOptions << TestConfig.OperatorString;
+  CompilerOptions << " -DOPERAND2=";
+  if (TestConfig.IsBinaryOp) {
+    CompilerOptions<< (TestConfig.IsScalarOp ? "LVOp.ScalarInput" : "LVOp.VecInput2");
+  }
+  CompilerOptions << " -DFUNC=";
+  CompilerOptions << TestConfig.IntrinsicString;
+  CompilerOptions << " -DCLAMP_ARGC=";
+  CompilerOptions << ClampArgC;
+  CompilerOptions << " -DCLAMP_ARGT=";
+  CompilerOptions << ClampArgT;
+
+  // We have to construct the string outside of the lambda. Otherwise it's
```

Comment:
The helper class really should be copying the string so this can't happen.

Reply:
You wrote that code though, so why not make it do that yourself? :)

Reply:
The st::ShaderOp should own the strings lifetime. But I didn't want to muck with that class. So instead, associating the strings lifetime with the test method seemed like the better option.